Add Observables functionality

cremebrule opened the pull request:
Adding observables functionality via the `Observable` class to robosuite to improve sim2real transferability.

This standardizes and modularizes how observations are computed and gathered within a given env. Now, all observations received from the `env.step()` call can be modified programmatically in either a deterministic or stochastic way. Sensor realism has been increased with added support for individual sensor sampling rates, corruption, delay, and filtering. The OTS behavior (obs dict structure, default no corruption / delay / filtering) has been preserved for backwards compatibility.

Each `Observable` owns its own `_sensor`, `_corrupter`, `_delayer`, and `_filter` functions, which are used to process new data computed during its `update()` call which is called after every _simulation_ timestep, NOT policy step! (Note, however that a new value is only computed once per sampling period, NOT at every `update()` call). Its functionality is described briefly described below:

- `_sensor`: Arbitrary function that takes in an observation cache and computes the "raw" (potentially ground truth) value for the given observable. It can potentially leverage pre-computed values from the observation cache to compute its output. The `@sensor` decorator is provided to denote this type of function, and guarantees a modality for this sensor as well.
- `_corrupter`: Arbitrary function that takes in the output from `_sensor` and outputs the corrupted data.
- `_delayer`: Arbitrary function that takes no arguments and outputs a float time value (in seconds), denoting how much delay should be applied to the next sampling cycle
- `_filter`: Arbitrary function that takes in the output of `_corrupter` and outputs the filtered data.

All of the above can either be (re-)defined at initialization or during runtime. Utility functions have been provided in the `base.py` environment module to easily interact with all observables owned by the environment.

Some standard corrupter and delayer function generators are provided ([deterministic / uniform / gaussian] [corruption / delay]), including dummy no-ops for standard functions. All of this can be found in `robosuite.utils.observables`, and has been heavily documented.

An example script demo'ing the new functionality can be found in `robosuite/demos/demo_sensor_corruption.py`.

yukezhu commented:
This is great work!

The `demo_sensor_corruption.py` script shows a use case of adding salt-and-pepper noise for image observations. Can we also include an example of sensor noise & delay for proprioception?

Also, `toggle_corruption_on_click` does not work in `demo_sensor_corruption.py`.

Also, can domain (dynamics) randomization be integrated part of the Observable functions (e.g., as a `corrupter`)? Or do they have to be implemented as wrappers?

The test `tests/test_environments/test_all_environments.py` seems to break with this new PR (actually, it seems to break in master as well).

cremebrule commented:
> This is great work!
> 
> The `demo_sensor_corruption.py` script shows a use case of adding salt-and-pepper noise for image observations. Can we also include an example of sensor noise & delay for proprioception?

Yep, added (see the printouts while running the demo)!

> Also, `toggle_corruption_on_click` does not work in `demo_sensor_corruption.py`.

Another user has brought this up before -- this is unfortunately a bad naming convention. The toggling is tied to the grasping action input, so the click on the spacemouse, OR the spacebar for the keyboard. This has been updated in both `demo_sensor_corruption.py` and `demo_device_control.py` from `XX-on-click` to `XX-on-grasp` for clarity.

> Also, can domain (dynamics) randomization be integrated part of the Observable functions (e.g., as a `corrupter`)? Or do they have to be implemented as wrappers?

Will dynamics randomization actually alter the simulation dynamics? If so, this would be abusing the observables functionality, since observables are intended to be passive components that do not impact the actual simulation. I was assuming this was the case and was planning to implement this as another `Modder` class that can be associated with the `DomainRandomization` wrapper we already have.

> The test `tests/test_environments/test_all_environments.py` seems to break with this new PR (actually, it seems to break in master as well).

Fixed these bugs, and verified that all tests now pass.

yukezhu commented:
Yes, dynamics randomization will make changes to the simulation dynamics, so it needs to be implemented as a `Modder`. Making it as part of the `DomainRandomization` wrapper sounds good. What's the desired behavior of `toggle_corruption_on_grasp`? The salt-and-pepper noise seems to appear with both open/closed grasps.

![image](https://user-images.githubusercontent.com/1192449/103321101-a9d82380-49fd-11eb-9596-450ee4c18d27.png)
![image](https://user-images.githubusercontent.com/1192449/103321111-b0ff3180-49fd-11eb-8ce4-6b4a965040d1.png)

cremebrule commented:
The toggling occurs once per grasping cycle (e.g., open -> close -> open) -- does the corruption not toggle if you press the spacebar multiple times (assuming you're using the keyboard device)?

I tried to think of a better naming convention but couldn't think of anything -- definitely open to renaming that for better clarity.

yukezhu commented:
I did try pressing the spacebar multiple times, and the corruption did not
toggle. This test was done on macOS Catalina. I think the new naming is
fine, as long as the functionality works.

>

cremebrule commented:
Hmmm, that's very strange. I just verified that it works locally on my machine (MacOS Mojave, Python 3.7) with the default keyboard device. When corruption is toggled off, there should be no corruption as shown below:

![image](https://user-images.githubusercontent.com/37633717/103324364-223de500-49fc-11eb-9879-a2074e119379.png)

Can you add a print statement at L231 of `demo_sensor_corruption.py` to make sure that code is being called every time you press spacebar? I can't think of a reason why we are having this behavior disparity.

yukezhu commented:
I see. I made some local edits for debugging, and forgot to revert them back. Now it works on my side! This PR looks good to me. Let's merge it once either @roberto-martinmartin or @amandlek is able to take a second look.

amandlek commented on tests/test_environments/test_all_environments.py:
```diff
@@ -2,7 +2,7 @@
 
 This runs some basic sanity checks on the environment, namely, checking that:
-    - robot-state exists in the obs, and is a flat array
+    - proprio-state exists in the obs, and is a flat array
```
is there a good reason for changing from "robot-state" to "proprio-state"?

cremebrule replied:
this change directly stemmed from the new `modality` attribute associated with a given `Observable` / `sensor` -- when the `obs_dict` is constructed after an `env.step()` call, individual keys are also concatenated together based on their associated sensor's `modality`.

So it seemed to me that defining the modality of the robot sensors as `proprio` vs. `robot` was more semantically meaningful and accurate, not to mention that the final modality keys `robot0_proprio-state` seems to be more informational vs what we had before (`robot0_robot-state`).

amandlek replied:
that's fair

amandlek reviewed (commented):
mostly LGTM, but left some comments

cremebrule commented on robosuite/environments/manipulation/nut_assembly.py:
```diff
@@ -448,6 +448,114 @@ def _get_reference(self):
+            # Immediately return default value if cache is empty
+            if any([name not in obs_cache for name in
+                    [f"{nut_name}_pos", f"{nut_name}_quat", "world_pose_in_gripper"]]):
+                return np.zeros(3)
```
This is to allow the `Observable` object that owns this sensor to be able to infer data shape information (e.g.: is this an array? or is this a scalar?), see https://github.com/ARISE-Initiative/robosuite/pull/146/files#diff-1e7090b536b9ec1f865375071b2667c19aad82b6250a147d25d21d8930df92c1R325

cremebrule commented on robosuite/environments/manipulation/nut_assembly.py:
```diff
@@ -468,83 +576,23 @@ def _reset_internal(self):
-            di["object-state"] = np.concatenate([di[k] for k in object_state_keys])
-
-        return di
+        # Make sure to update sensors' active and enabled states
```
Enabled = The observable will internally actually update its value and add its result to the `obs_cache`.

Active = The observable output will actually be returned in the obs_dict from `_get_observations()`.

The main distinction is that some observables should be internal but not necessarily required / desired for the actual outputted observation dictionary. A concrete use case is the `joint_pos` observable vs. the `joint_pos_cos`, `joint_pos_sin` observables. The former is enabled but NOT active since its outputted value is required to calculate the corresponding cos / sin values, but not desired for the actual outputted `obs_dict` since it contains redundant information relative to the cos / sin values.

cremebrule replied:
see https://github.com/ARISE-Initiative/robosuite/pull/146/files#diff-1e7090b536b9ec1f865375071b2667c19aad82b6250a147d25d21d8930df92c1R238 and https://github.com/ARISE-Initiative/robosuite/pull/146/files#diff-1e7090b536b9ec1f865375071b2667c19aad82b6250a147d25d21d8930df92c1R250

cremebrule commented on robosuite/robots/single_arm.py:
```diff
@@ -8,7 +8,8 @@
 
 from robosuite.robots.manipulator import Manipulator
-from robosuite.utils.control_utils import DeltaBuffer, RingBuffer
+from robosuite.utils.buffers import DeltaBuffer, RingBuffer
```
I believe these are being used -- it should only be in `single_arm` and `bimanual` right (and maybe the controllers as well). It's for keeping track of recent values, e.g. for computing recent ee_torques and such.

amandlek replied:
my bad, it was removed and added back - so I thought it was a new change - but this was probably due to some formatting stuff

cremebrule commented on robosuite/utils/observables.py:
```diff
@@ -0,0 +1,348 @@
+    """
+    Decorator that should be added to any sensors that will be an observable.
+
+    Decorated functions should have signature:
```
great point -- done! `modality` is effectively an attribute associated with the function callable that the decorator wraps. It should be a string value, and the `modality=modality` you see is just referencing a local variable that was defined earlier so I didn't have to hard code the string value for sensors sharing the same modality

cremebrule replied:
e.g.: https://github.com/ARISE-Initiative/robosuite/pull/146/files#diff-ea79ecdf14c6f9d19efd297841abe5a7c4f5b2f7863dc14bdcbe0350263dc2cdR325

amandlek replied:
makes sense!

amandlek commented on robosuite/utils/observables.py:
```diff
@@ -9,6 +9,18 @@ def sensor(modality):
+        @sensor(modality="proprio")
+        def joint_pos(obs_cache):
+            # Always handle case if obs_cache is empty
+            if not obs_cache:
```
do you mean to check if obs_cache is an empty dictionary, or if it's None? in either case, I prefer to use checks like len(obs_cache) > 0 or obs_cache is not None - there can be tricky bugs by directly checking boolean values

cremebrule replied:
Good point -- I think I originally designed the interface to be able to handle either case, though I think the only (current) use case passes in an empty dict
